keras_fsl/utils/datasets.py

```diff
@@ -0,0 +1,121 @@
+
+    return annotations_mapper
+
+
+def read_decode_and_crop_jpeg(annotation: TENSOR_MAP) -> TF_TENSOR:
+    """
+    Args:
+        annotation (TENSOR_MAP): contains 'filename' (path to the image) and possibly 'crop_window' (box coordinates)
+    Returns:
+        TF_TENSOR: the crop described by annotations as a uint8 array
+    """
+    image_content = tf.io.read_file(filename=annotation["filename"])
+    if "crop_window" in annotation:
+        return tf.io.decode_and_crop_jpeg(image_content, crop_window=annotation["crop_window"], channels=3)
+
+    return tf.io.decode_jpeg(image_content, channels=3)
+
+
+def cache_with_tf_record(filename: Union[str, pathlib.Path]) -> Callable[[tf.data.Dataset], tf.data.TFRecordDataset]:
+    """
+    Similar to tf.data.Dataset.cache but writes a tf record file instead. Compared to base .cache method, it also insures that the whole
+    dataset is cached
+    """
+
+    def _cache(dataset):
+        if not isinstance(dataset.element_spec, dict):
+            raise ValueError(f"dataset.element_spec should be a dict but is {type(dataset.element_spec)} instead")
+        with tf.io.TFRecordWriter(str(filename)) as writer:
+            for sample in dataset.map(transform(**{name: tf.io.serialize_tensor for name in dataset.element_spec.keys()})):
+                writer.write(
+                    tf.train.Example(
+                        features=tf.train.Features(
+                            feature={
+                                key: tf.train.Feature(bytes_list=tf.train.BytesList(value=[value.numpy()]))
+                                for key, value in sample.items()
+                            }
+                        )
+                    ).SerializeToString()
+                )
+        return (
+            tf.data.TFRecordDataset(str(filename), num_parallel_reads=tf.data.experimental.AUTOTUNE)
+            .map(
+                partial(
+                    tf.io.parse_single_example,
+                    features={name: tf.io.FixedLenFeature((), tf.string) for name in dataset.element_spec.keys()},
+                ),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
+            .map(
+                transform(
+                    **{name: partial(tf.io.parse_tensor, out_type=spec.dtype) for name, spec in dataset.element_spec.items()}
+                )
+            )
+        )
+
+    return _cache
+
+
+def clear_cache(filename):
+    """
+    Clear cache created with tf.data.Dataset.cache given the name used for cache creation:
+        e.g. dataset.cache(filename) will produce filename.index and filename.data-xxx files
+    Args:
+        filename (Union[str, pathlib.Path]): filename used during cache creation
+
+    Returns:
+        List[pathlib.Path]: list of deleted items
+    """
+    filename = Path(filename)
+    files = list(filename.parent.glob(f"{filename.name}.*"))
+    for file in files:
+        file.unlink()
```

Comment:
c'est pas bizarre du coup de donner un filename qui n'est pas vraiment pris en compte ? il faudrait pas plutôt avoir un cache dir ?

Reply:
c'est le truc avec `tf.data.Dataset.cache` le paramètre s'appelle `filename` mais en vrai ça crée un filename.index et des shard https://www.tensorflow.org/api_docs/python/tf/data/Dataset#cache

Reply:
oui, c'est "assez" classique. Tu proposes de reproduire pour avoir la cohérence ?

Reply:
j'ai voulu garder le même comportement parce que je voulais juste wrap `tf.data.Dataset.cache` pour itérer et ajouter le clear